src/vs/editor/contrib/codeAction/browser/codeActionMenu.ts

```diff
@@ -50,34 +63,297 @@ export interface CodeActionShowOptions {
 	readonly fromLightbulb?: boolean;
 }
+export interface ICodeActionMenuItem {
+	title: string;
+	detail: string;
+	action: IAction;
+	decoratorRight?: string;
+	isSeparator?: boolean;
+	isEnabled: boolean;
+	index: number;
+	disposables?: IDisposable[];
+}
+
+export interface ICodeMenuOptions {
+	useCustomDrawn?: boolean;
+	ariaLabel?: string;
+	ariaDescription?: string;
+	minBottomMargin?: number;
+	optionsAsChildren?: boolean;
+}
 
-export class CodeActionMenu extends Disposable {
+export interface ICodeActionMenuTemplateData {
+	root: HTMLElement;
+	text: HTMLElement;
+	detail: HTMLElement;
+	decoratorRight: HTMLElement;
+	disposables: IDisposable[];
+}
 
-	private _visible: boolean = false;
+const TEMPLATE_ID = 'codeActionWidget';
+class CodeMenuRenderer implements IListRenderer<ICodeActionMenuItem, ICodeActionMenuTemplateData> {
+	get templateId(): string { return TEMPLATE_ID; }
+
+	renderTemplate(container: HTMLElement): ICodeActionMenuTemplateData {
+		const data: ICodeActionMenuTemplateData = Object.create(null);
+		data.disposables = [];
+		data.root = container;
+		data.text = document.createElement('span');
+		// data.detail = document.createElement('');
+		container.append(data.text);
+		// container.append(data.detail);
+
+		return data;
+	}
+	renderElement(element: ICodeActionMenuItem, index: number, templateData: ICodeActionMenuTemplateData): void {
+		const data: ICodeActionMenuTemplateData = templateData;
+
+		const text = element.title;
+		// const detail = element.detail;
+
+		const isEnabled = element.isEnabled;
+		const isSeparator = element.isSeparator;
+
+		data.text.textContent = text;
+		// data.detail.textContent = detail;
+
+		if (!isEnabled) {
+			data.root.classList.add('option-disabled');
+			data.root.style.backgroundColor = 'transparent !important';
+			data.root.style.color = 'rgb(204, 204, 204, 0.5)';
+			data.root.style.cursor = 'default';
+		} else {
+			data.root.classList.remove('option-disabled');
+		}
+
+		if (isSeparator) {
+			data.root.classList.add('separator');
+			data.root.style.height = '10px';
+		}
+
+	}
+	disposeTemplate(templateData: ICodeActionMenuTemplateData): void {
+		templateData.disposables = dispose(templateData.disposables);
+	}
+}
+
+export class CodeActionMenu extends Disposable implements IEditorContribution {
+	private readonly _ctxMenuWidgetIsFocused?: IContextKey<boolean>;
+	private readonly editor: ICodeEditor;
 	private readonly _showingActions = this._register(new MutableDisposable<CodeActionSet>());
+	private readonly _disposables = new DisposableStore();
+	private readonly _onDidHideContextMenu = new Emitter<void>();
+	private codeActionList!: List<ICodeActionMenuItem>;
+	private options: ICodeActionMenuItem[] = [];
+	private _visible: boolean = false;
+	readonly onDidHideContextMenu = this._onDidHideContextMenu.event;
+	private _ctxMenuWidgetVisible!: IContextKey<boolean>;
+	private viewItems: ICodeActionMenuItem[] = [];
+	private focusedEnabledItem!: number;
```

Comment:
Except for exceptional cases, try to avoid using `!` and instead mark the types as `type | undefined`